Parsers such as separated can collect into Vec, but not HashSet or BTreeSet

### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### winnow version

0.5.19

### Describe your use case

For an advent of code puzzle I wanted to directly parse a list of numbers into a `HashSet`

```rust
 fn parse_numbers(input: &mut &str) -> PResult<HashSet<u64>> {
        let numbers = separated(1.., dec_uint, multispace1).parse_next(input)?;
        PResult::Ok(numbers)
}
```

```
error[E0277]: the trait bound `std::collections::HashSet<u64>: winnow::stream::Accumulate<_>` is not satisfied
   --> src/days/day4.rs:88:23
    |
42  |         let numbers = separated(1.., dec_uint, multispace1).context(StrContext::Label("parse numbers")).parse_next(input)?;
    |                       ^^^^^^^^^ the trait `winnow::stream::Accumulate<_>` is not implemented for `std::collections::HashSet<u64>`
    |
    = help: the following other types implement trait `winnow::stream::Accumulate<T>`:
              <usize as winnow::stream::Accumulate<T>>
              <std::collections::HashMap<K, V, S> as winnow::stream::Accumulate<(K, V)>>
              <std::collections::BTreeMap<K, V> as winnow::stream::Accumulate<(K, V)>>
              <std::vec::Vec<T> as winnow::stream::Accumulate<T>>
              <std::vec::Vec<T> as winnow::stream::Accumulate<&'i [T]>>
              <std::string::String as winnow::stream::Accumulate<char>>
              <std::string::String as winnow::stream::Accumulate<&'i str>>
              <() as winnow::stream::Accumulate<T>>
note: required by a bound in `winnow::combinator::separated`
   --> /home/daniel/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winnow-0.5.21/src/combinator/multi.rs:447:8
    |
440 | pub fn separated<I, O, C, O2, E, P, S>(
    |        --------- required by a bound in this function
...
447 |     C: Accumulate<O>,
    |        ^^^^^^^^^^^^^ required by this bound in `separated`
```


### Describe the solution you'd like

The code should simply compile and return a HashSet/BTreeSet.

### Alternatives, if applicable

Changing the return value to PResult<Vec<_>> and then converting that to a HashSet would work, but I think it would be useful to "accumulate" directly into a HashSet and BTreeSet.

### Additional Context

_No response_

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -26,8 +26,12 @@ use crate::error::ErrMode;
 
 #[cfg(feature = "alloc")]
 use crate::lib::std::collections::BTreeMap;
+#[cfg(feature = "alloc")]
+use crate::lib::std::collections::BTreeSet;
 #[cfg(feature = "std")]
 use crate::lib::std::collections::HashMap;
+#[cfg(feature = "std")]
+use crate::lib::std::collections::HashSet;
 #[cfg(feature = "alloc")]
 use crate::lib::std::string::String;
 #[cfg(feature = "alloc")]
@@ -2458,6 +2462,41 @@ where
     #[inline(always)]
     fn accumulate(&mut self, (key, value): (K, V)) {
         self.insert(key, value);
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<K> Accumulate<K> for BTreeSet<K>
+where
+    K: crate::lib::std::cmp::Ord,
+{
+    #[inline(always)]
+    fn initial(_capacity: Option<usize>) -> Self {
+        BTreeSet::new()
+    }
+    #[inline(always)]
+    fn accumulate(&mut self, key: K) {
+        self.insert(key);
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K, S> Accumulate<K> for HashSet<K, S>
+where
+    K: crate::lib::std::cmp::Eq + crate::lib::std::hash::Hash,
+    S: BuildHasher + Default,
+{
+    #[inline(always)]
+    fn initial(capacity: Option<usize>) -> Self {
+        let h = S::default();
+        match capacity {
+            Some(capacity) => HashSet::with_capacity_and_hasher(clamp_capacity::<K>(capacity), h),
+            None => HashSet::with_hasher(h),
+        }
+    }
+    #[inline(always)]
+    fn accumulate(&mut self, key: K) {
+        self.insert(key);
     }
 }
 

